src/include/buffer/buffer_pool_manager.h

```diff
@@ -17,20 +17,117 @@
+    // 0-FetchPage  1-UnpinPage  2-FlushPage 3-NewPage 4-DeletePage
+    // 5-FlushAllPages
+    static const int num_types = 6;
+    int counts[num_types];
+
+    void Reset() {
+      for (int &count : counts) {
+        count = 0;
+      }
+    }
+
+    void AddCount(FuncType func_type) { ++counts[func_type]; }
+
+    // Make sure fetch page function only calls fetch page once and
+    // does not call other functions
+    void CheckFetchPage() {
+      EXPECT_EQ(counts[0], 1) << "has to call FetchPage once";
+      for (int i = 1; i < num_types; ++i) {
+        EXPECT_EQ(counts[i], 0) << "FetchPage Should not call other functions";
+      }
+    }
+
+    void CheckUnpinPage() {
+      EXPECT_EQ(counts[1], 1) << "has to call UnpinPage once";
+      for (int i = 0; i < num_types; ++i) {
+        if (i != 1) {
+          EXPECT_EQ(counts[i], 0) << "UnPinPage Should not call other functions";
+        }
+      }
+    }
+
+    void CheckFlushPage() {
+      EXPECT_EQ(counts[2], 1) << "has to call FlushPage once";
+      for (int i = 0; i < num_types; ++i) {
+        if (i != 2) {
+          EXPECT_EQ(counts[i], 0) << "FlushPage Should not call other functions";
+        }
+      }
+    }
+
+    void CheckNewPage() {
+      EXPECT_EQ(counts[3], 1) << "has to call NewPage once";
+      for (int i = 0; i < num_types; ++i) {
+        if (i != 3) {
+          EXPECT_EQ(counts[i], 0) << "NewPage Should not call other functions";
+        }
+      }
+    }
+
+    void CheckDeletePage() {
+      EXPECT_EQ(counts[4], 1) << "has to call DeletePage once";
+      for (int i = 0; i < num_types; ++i) {
+        if (i != 4) {
+          EXPECT_EQ(counts[i], 0) << "DeletePage Should not call other functions";
+        }
+      }
+    }
+
+    void CheckFlushAllPages() {
+      for (int i = 1; i < 5; ++i) {
+        EXPECT_EQ(counts[i], 0) << "FlushAllPage Should not call other functions";
+      }
+      EXPECT_EQ(counts[5], 1) << "has to call FlushAllPage once";
+    }
+  };
+
  public:
   enum class CallbackType { BEFORE, AFTER };
-  using bufferpool_callback_fn = void (*)(enum CallbackType, const page_id_t page_id);
+  using bufferpool_callback_fn = void (BufferPoolManager::*)(enum CallbackType type, FuncType func_type);
 
+  void counter_callback(enum CallbackType type, FuncType func_type) {
+    if (type == CallbackType::BEFORE) {
+      counter.Reset();
+    } else {
+      switch (func_type) {
+        case FuncType::FetchPage:
+          counter.CheckFetchPage();
+          break;
+        case FuncType::UnpinPage:
+          counter.CheckUnpinPage();
+          break;
+        case FuncType::FlushPage:
+          counter.CheckFlushPage();
+          break;
+        case FuncType::NewPage:
+          counter.CheckNewPage();
+          break;
+        case FuncType::DeletePage:
+          counter.CheckDeletePage();
+          break;
+        case FuncType::FlushAllPages:
+          counter.CheckFlushAllPages();
+          break;
+      }
+    }
+  }
```

Comment:
Move the counter callback to be outside of the class. Define it in a separate file. It is also only used for testing, so we can potentially move into the `test` directory.

We want to reduce the amount of code in the parts of the system that the students have to look at to minimize the complexity.